File: crates/ressrf-core/src/cidr.rs
use alloc::string::String;
use alloc::vec::Vec;
use core::fmt;
use core::net::{IpAddr, Ipv4Addr, Ipv6Addr};

use crate::error::{DataTier, Error};

/// A parsed CIDR range with precomputed network bytes and mask for fast containment checks.
#[derive(Clone, PartialEq, Eq, Hash)]
pub struct Cidr {
    /// Normalized to IPv6 (IPv4 mapped to `::ffff:0:0/96` + prefix).
    network: [u8; 16],
    /// Prefix length in the IPv6 address space (0..=128).
    prefix_len: u8,
    /// Precomputed mask bytes for bitwise containment.
    mask: [u8; 16],
    /// Original string representation for display/audit.
    original: String,
    /// Which data tier this CIDR came from.
    source: DataTier,
}

impl fmt::Debug for Cidr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Cidr")
            .field("original", &self.original)
            .field("prefix_len", &self.prefix_len)
            .field("source", &self.source)
            .finish_non_exhaustive()
    }
}

impl fmt::Display for Cidr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.original)
    }
}

impl Cidr {
    /// Parse a CIDR string (e.g. `192.168.0.0/16` or `fe80::/10`).
    ///
    /// Rejects:
    /// - Octal IP octets (leading zeros like `0177.0.0.1`)
    /// - Hex IP octets (like `0x7f.0.0.1`)
    /// - IPv6 zone IDs (like `fe80::1%eth0/64`)
    /// - Host bits set (the network portion must be correctly masked)
    pub fn parse(s: &str, source: DataTier) -> crate::Result<Self> {
        let s = s.trim();

        // Reject zone IDs
        if s.contains('%') {
            return Err(Error::Parse("zone IDs not allowed in CIDR notation".into()));
        }

        let (addr_str, prefix_str) = s
            .rsplit_once('/')
            .ok_or_else(|| Error::Parse("missing '/' in CIDR notation".into()))?;

        let prefix_len: u8 = prefix_str
            .parse()
            .map_err(|_| Error::Parse("invalid prefix length".into()))?;

        let ip = parse_ip_strict(addr_str)?;

        let (normalized_ip, adjusted_prefix) = normalize_to_ipv6(ip, prefix_len)?;

        let network_bytes = normalized_ip.octets();
        let mask = compute_mask(adjusted_prefix);

        // Verify network address: all host bits must be zero
        for i in 0..16 {
            if network_bytes[i] & !mask[i] != 0 {
                return Err(Error::Parse("host bits set in network address".into()));
            }
        }

        Ok(Self {
            network: network_bytes,
            prefix_len: adjusted_prefix,
            mask,
            original: String::from(s),
            source,
        })
    }

    /// Parse a CIDR, allowing host bits (auto-masks to network address).
    /// Used for user-supplied ranges where strictness is less important than usability.
    pub fn parse_loose(s: &str, source: DataTier) -> crate::Result<Self> {
        let s = s.trim();

        if s.contains('%') {
            return Err(Error::Parse("zone IDs not allowed in CIDR notation".into()));
        }

        let (addr_str, prefix_str) = s
            .rsplit_once('/')
            .ok_or_else(|| Error::Parse("missing '/' in CIDR notation".into()))?;

        let prefix_len: u8 = prefix_str
            .parse()
            .map_err(|_| Error::Parse("invalid prefix length".into()))?;

        let ip = parse_ip_strict(addr_str)?;
        let (normalized_ip, adjusted_prefix) = normalize_to_ipv6(ip, prefix_len)?;

        let mut network_bytes = normalized_ip.octets();
        let mask = compute_mask(adjusted_prefix);

        // Zero out host bits
        for i in 0..16 {
            network_bytes[i] &= mask[i];
        }

        Ok(Self {
            network: network_bytes,
            prefix_len: adjusted_prefix,
            mask,
            original: String::from(s),
            source,
        })
    }

    /// Check if an IP address is contained in this CIDR range.
    #[must_use]
    pub fn contains(&self, ip: IpAddr) -> bool {
        let normalized = match ip {
            IpAddr::V4(v4) => v4.to_ipv6_mapped(),
            IpAddr::V6(v6) => v6,
        };
        let octets = normalized.octets();

        octets
            .iter()
            .zip(self.mask.iter())
            .zip(self.network.iter())
            .all(|((&octet, &mask), &net)| (octet & mask) == net)
    }

    #[must_use]
    pub fn prefix_len(&self) -> u8 {
        self.prefix_len
    }

    #[must_use]
    pub fn source(&self) -> DataTier {
        self.source
    }

    #[must_use]
    pub fn original(&self) -> &str {
        &self.original
    }
}

/// A collection of CIDR ranges optimized for containment lookups.
#[derive(Debug, Clone, Default)]
pub struct CidrSet {
    ranges: Vec<Cidr>,
}

impl CidrSet {
    #[must_use]
    pub fn new() -> Self {
        Self { ranges: Vec::new() }
    }

    pub fn add(&mut self, cidr: Cidr) {
        self.ranges.push(cidr);
    }

    /// Check if any range in the set contains the given IP.
    /// Returns the first matching CIDR if found.
    #[must_use]
    pub fn contains(&self, ip: IpAddr) -> Option<&Cidr> {
        self.ranges.iter().find(|cidr| cidr.contains(ip))
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.ranges.is_empty()
    }

    #[must_use]
    pub fn len(&self) -> usize {
        self.ranges.len()
    }
}

/// Parse an IP address strictly: reject octal (leading zeros) and hex notation.
fn parse_ip_strict(s: &str) -> crate::Result<IpAddr> {
    let s = s.trim();

    // Try IPv6 first (contains ':')
    if s.contains(':') {
        let v6: Ipv6Addr = s
            .parse()
            .map_err(|e| Error::Parse(alloc::format!("invalid IPv6 address: {e}")))?;
        return Ok(IpAddr::V6(v6));
    }

    // IPv4: check for octal/hex octets
    reject_ambiguous_ipv4(s)?;

    let v4: Ipv4Addr = s
        .parse()
        .map_err(|e| Error::Parse(alloc::format!("invalid IPv4 address: {e}")))?;
    Ok(IpAddr::V4(v4))
}

/// Reject IPv4 addresses with leading zeros (octal) or hex prefixes.
fn reject_ambiguous_ipv4(s: &str) -> crate::Result<()> {
    for octet_str in s.split('.') {
        if octet_str.is_empty() {
            return Err(Error::Parse("empty octet in IPv4 address".into()));
        }
        // Reject hex: "0x..." or "0X..."
        if octet_str.starts_with("0x") || octet_str.starts_with("0X") {
            return Err(Error::Parse(
                "hex notation not allowed in IP addresses".into(),
            ));
        }
        // Reject octal: leading zero followed by more digits (e.g. "0177", "010")
        if octet_str.len() > 1 && octet_str.starts_with('0') {
            return Err(Error::Parse(
                "leading zeros (octal notation) not allowed in IP addresses".into(),
            ));
        }
    }
    Ok(())
}

/// Normalize an IP to IPv6, adjusting the prefix length accordingly.
/// IPv4 addresses get mapped to `::ffff:a.b.c.d` with prefix += 96.
fn normalize_to_ipv6(ip: IpAddr, prefix_len: u8) -> crate::Result<(Ipv6Addr, u8)> {
    match ip {
        IpAddr::V4(v4) => {
            if prefix_len > 32 {
                return Err(Error::Parse("IPv4 prefix length must be 0..=32".into()));
            }
            let mapped = v4.to_ipv6_mapped();
            let adjusted = prefix_len + 96;
            Ok((mapped, adjusted))
        }
        IpAddr::V6(v6) => {
            if prefix_len > 128 {
                return Err(Error::Parse("IPv6 prefix length must be 0..=128".into()));
            }
            Ok((v6, prefix_len))
        }
    }
}

/// Compute a 128-bit mask from a prefix length.
fn compute_mask(prefix_len: u8) -> [u8; 16] {
    let mut mask = [0u8; 16];
    let full_bytes = (prefix_len / 8) as usize;
    let remaining_bits = prefix_len % 8;

    for byte in mask.iter_mut().take(full_bytes) {
        *byte = 0xFF;
    }
    if full_bytes < 16 && remaining_bits > 0 {
        mask[full_bytes] = 0xFF << (8 - remaining_bits);
    }
    mask
}

/// Parse an IP address strictly (for use in policy checks on user-supplied IPs).
/// Exported so the URI validator can reuse it.
pub fn parse_ip(s: &str) -> crate::Result<IpAddr> {
    parse_ip_strict(s)
}

/// Check if a string looks like a bare IP literal (v4 or v6, with optional brackets).
///
/// Uses the strict parser, so ambiguous IPv4 forms (octal `0177.0.0.1`, hex
/// `0x7f000001`, decimal `2130706433`, shorthand `127.1`) return `false` here
/// to avoid silently letting them pass through. Use [`is_ambiguous_ip`] to
/// detect those forms separately.
#[must_use]
pub fn is_ip_literal(s: &str) -> bool {
    let s = s.trim();
    // Bracketed IPv6: [::1]
    let inner = if s.starts_with('[') && s.ends_with(']') {
        &s[1..s.len() - 1]
    } else {
        s
    };
    parse_ip_strict(inner).is_ok()
}

/// Detect host strings that look like an IP address but use ambiguous or
/// non-canonical encodings (decimal integer, octal, hex, shorthand).
///
/// These forms are rejected by [`parse_ip_strict`] and would otherwise be
/// silently treated as DNS hostnames. Examples:
/// - `2130706433` (decimal `127.0.0.1`)
/// - `0177.0.0.1` (octal `127.0.0.1`)
/// - `0x7f000001` or `0x7f.0.0.1` (hex `127.0.0.1`)
/// - `127.1` or `127.0.1` (shorthand `127.0.0.1`)
///
/// Returns `false` for genuine hostnames (`example.com`, `localhost`).
/// Returns `false` for canonical IPv4/IPv6 (use [`is_ip_literal`] for those).
#[must_use]
pub fn is_ambiguous_ip(s: &str) -> bool {
    let s = s.trim();

    // Strip IPv6 brackets if present: bracketed forms are unambiguous IPv6.
    if s.starts_with('[') && s.ends_with(']') {
        return false;
    }

    if s.is_empty() {
        return false;
    }

    // IPv6 (contains ':') is unambiguous: either parses or doesn't.
    if s.contains(':') {
        return false;
    }

    // Already-canonical IPv4 (4 dotted decimal octets) is unambiguous.
    if parse_ip_strict(s).is_ok() {
        return false;
    }

    // Hex IP (`0xNN.0xNN.0xNN.0xNN` or single integer like `0x7f000001`)
    let lower = s.to_ascii_lowercase();
    if lower.starts_with("0x")
        && lower
            .trim_start_matches("0x")
            .chars()
            .all(|c| c.is_ascii_hexdigit())
    {
        return true;
    }
    for label in s.split('.') {
        let lower_label = label.to_ascii_lowercase();
        if lower_label.starts_with("0x")
            && lower_label
                .trim_start_matches("0x")
                .chars()
                .all(|c| c.is_ascii_hexdigit())
        {
            return true;
        }
    }

    // Decimal integer form: all digits, no dots (e.g. `2130706433`).
    if !s.is_empty() && s.chars().all(|c| c.is_ascii_digit()) {
        return true;
    }

    // Dotted form with all-numeric labels.
    if s.contains('.') {
        let mut all_numeric = true;
        let mut octet_count = 0;
        let mut has_octal = false;
        for label in s.split('.') {
            octet_count += 1;
            if label.is_empty() {
                all_numeric = false;
                break;
            }
            if !label.chars().all(|c| c.is_ascii_digit()) {
                all_numeric = false;
                break;
            }
            // Leading-zero octets are octal (e.g. `0177`).
            if label.len() > 1 && label.starts_with('0') {
                has_octal = true;
            }
        }

        if all_numeric {
            // Octal form (e.g. `0177.0.0.1`) -- non-canonical.
            if has_octal {
                return true;
            }
            // Shorthand IPv4 (e.g. `127.1`, `127.0.1`) -- non-canonical.
            if octet_count != 4 {
                return true;
            }
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_ipv4_cidr() {
        let c = Cidr::parse("192.168.0.0/16", DataTier::Iana).unwrap();
        assert!(c.contains("192.168.1.1".parse().unwrap()));
        assert!(c.contains("192.168.255.255".parse().unwrap()));
        assert!(!c.contains("192.169.0.0".parse().unwrap()));
        assert!(!c.contains("10.0.0.1".parse().unwrap()));
    }

    #[test]
    fn parse_ipv6_cidr() {
        let c = Cidr::parse("fe80::/10", DataTier::Iana).unwrap();
        assert!(c.contains("fe80::1".parse().unwrap()));
        assert!(c.contains("febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff".parse().unwrap()));
        assert!(!c.contains("ff00::1".parse().unwrap()));
    }

    #[test]
    fn ipv4_mapped_ipv6_containment() {
        // 192.168.0.0/16 should match its IPv4-mapped IPv6 form
        let c = Cidr::parse("192.168.0.0/16", DataTier::Iana).unwrap();
        let mapped: IpAddr = "::ffff:192.168.1.1".parse().unwrap();
        assert!(c.contains(mapped));
    }

    #[test]
    fn ipv6_cidr_matches_mapped_ipv4() {
        // ::ffff:0.0.0.0/96 should match any IPv4 address (since all IPv4 maps into this)
        let c = Cidr::parse("::ffff:0.0.0.0/96", DataTier::Iana).unwrap();
        assert!(c.contains("1.2.3.4".parse().unwrap()));
        assert!(c.contains("192.168.0.1".parse().unwrap()));
    }

    #[test]
    fn reject_octal_ip() {
        let result = Cidr::parse("0177.0.0.1/32", DataTier::Iana);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("octal"));
    }

    #[test]
    fn reject_hex_ip() {
        let result = Cidr::parse("0x7f.0.0.1/32", DataTier::Iana);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("hex"));
    }

    #[test]
    fn reject_zone_id() {
        let result = Cidr::parse("fe80::1%eth0/64", DataTier::Iana);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("zone"));
    }

    #[test]
    fn slash_zero_matches_everything() {
        let c = Cidr::parse("0.0.0.0/0", DataTier::Iana).unwrap();
        assert!(c.contains("1.2.3.4".parse().unwrap()));
        assert!(c.contains("255.255.255.255".parse().unwrap()));
        // /0 in IPv4 means /96 in normalized space, so it matches all IPv4
        assert!(c.contains("192.168.0.1".parse().unwrap()));
    }

    #[test]
    fn slash_128_single_host() {
        let c = Cidr::parse("::1/128", DataTier::Iana).unwrap();
        assert!(c.contains("::1".parse().unwrap()));
        assert!(!c.contains("::2".parse().unwrap()));
    }

    #[test]
    fn slash_32_single_ipv4_host() {
        let c = Cidr::parse("169.254.169.254/32", DataTier::CspMetadata).unwrap();
        assert!(c.contains("169.254.169.254".parse().unwrap()));
        assert!(!c.contains("169.254.169.253".parse().unwrap()));
        assert!(!c.contains("169.254.169.255".parse().unwrap()));
    }

    #[test]
    fn host_bits_set_rejected() {
        let result = Cidr::parse("192.168.1.1/16", DataTier::Iana);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("host bits"));
    }

    #[test]
    fn parse_loose_allows_host_bits() {
        let c = Cidr::parse_loose("192.168.1.1/16", DataTier::UserDeny).unwrap();
        assert!(c.contains("192.168.0.1".parse().unwrap()));
        assert!(c.contains("192.168.255.255".parse().unwrap()));
    }

    #[test]
    fn ipv4_prefix_too_large() {
        let result = Cidr::parse("10.0.0.0/33", DataTier::Iana);
        assert!(result.is_err());
    }

    #[test]
    fn ipv6_prefix_too_large() {
        let result = Cidr::parse("::1/129", DataTier::Iana);
        assert!(result.is_err());
    }

    #[test]
    fn cidr_set_contains() {
        let mut set = CidrSet::new();
        set.add(Cidr::parse("10.0.0.0/8", DataTier::Iana).unwrap());
        set.add(Cidr::parse("172.16.0.0/12", DataTier::Iana).unwrap());
        set.add(Cidr::parse("192.168.0.0/16", DataTier::Iana).unwrap());

        assert!(set.contains("10.0.0.1".parse().unwrap()).is_some());
        assert!(set.contains("172.16.0.1".parse().unwrap()).is_some());
        assert!(set.contains("192.168.0.1".parse().unwrap()).is_some());
        assert!(set.contains("8.8.8.8".parse().unwrap()).is_none());
    }

    #[test]
    fn imds_ranges() {
        let aws_imds = Cidr::parse("169.254.169.254/32", DataTier::CspMetadata).unwrap();
        let aws_ecs = Cidr::parse("169.254.170.2/32", DataTier::CspMetadata).unwrap();
        let azure_wire = Cidr::parse("168.63.129.16/32", DataTier::CspMetadata).unwrap();

        assert!(aws_imds.contains("169.254.169.254".parse().unwrap()));
        assert!(!aws_imds.contains("169.254.169.253".parse().unwrap()));
        assert!(aws_ecs.contains("169.254.170.2".parse().unwrap()));
        assert!(azure_wire.contains("168.63.129.16".parse().unwrap()));
    }

    #[test]
    fn nat64_ranges() {
        let nat64_wk = Cidr::parse("64:ff9b::/96", DataTier::Override).unwrap();
        let nat64_local = Cidr::parse("64:ff9b:1::/48", DataTier::Override).unwrap();

        // NAT64 well-known wraps IPv4 in the low 32 bits
        assert!(nat64_wk.contains("64:ff9b::1.2.3.4".parse().unwrap()));
        assert!(nat64_wk.contains("64:ff9b::c0a8:0001".parse().unwrap()));

        // NAT64 local-use /48
        assert!(nat64_local.contains("64:ff9b:1::1".parse().unwrap()));
        assert!(!nat64_local.contains("64:ff9b:2::1".parse().unwrap()));
    }

    #[test]
    fn six_to_four_range() {
        let six_to_four = Cidr::parse("2002::/16", DataTier::Override).unwrap();
        assert!(six_to_four.contains("2002:c0a8:1::1".parse().unwrap()));
        assert!(!six_to_four.contains("2003::1".parse().unwrap()));
    }

    #[test]
    fn teredo_range() {
        let teredo = Cidr::parse("2001::/32", DataTier::Override).unwrap();
        assert!(teredo.contains("2001::1".parse().unwrap()));
        assert!(teredo.contains("2001:0000:ffff:ffff:ffff:ffff:ffff:ffff".parse().unwrap()));
        assert!(!teredo.contains("2001:0001::1".parse().unwrap()));
    }

    #[test]
    fn is_ip_literal_detection() {
        assert!(is_ip_literal("1.2.3.4"));
        assert!(is_ip_literal("::1"));
        assert!(is_ip_literal("[::1]"));
        assert!(!is_ip_literal("example.com"));
        assert!(!is_ip_literal("not-an-ip"));
    }

    #[test]
    fn is_ip_literal_rejects_ambiguous_forms() {
        // Strict parser rejects octal/hex/shorthand/decimal-integer.
        assert!(!is_ip_literal("0177.0.0.1"));
        assert!(!is_ip_literal("0x7f000001"));
        assert!(!is_ip_literal("0x7f.0.0.1"));
        assert!(!is_ip_literal("2130706433"));
        assert!(!is_ip_literal("127.1"));
    }

    #[test]
    fn is_ambiguous_ip_detection() {
        // Decimal integer form
        assert!(is_ambiguous_ip("2130706433"));
        assert!(is_ambiguous_ip("2852039166"));
        assert!(is_ambiguous_ip("0"));

        // Octal form
        assert!(is_ambiguous_ip("0177.0.0.1"));
        assert!(is_ambiguous_ip("0177.0000.0000.0001"));
        assert!(is_ambiguous_ip("017700000001"));

        // Hex form
        assert!(is_ambiguous_ip("0x7f000001"));
        assert!(is_ambiguous_ip("0xa9fea9fe"));
        assert!(is_ambiguous_ip("0x7f.0x0.0x0.0x1"));
        assert!(is_ambiguous_ip("0X7F000001"));

        // Shorthand form
        assert!(is_ambiguous_ip("127.1"));
        assert!(is_ambiguous_ip("127.0.1"));

        // Mixed (not canonical 4-dot form, hex prefix)
        assert!(is_ambiguous_ip("0x7f.0.0.1"));
    }

    #[test]
    fn is_ambiguous_ip_excludes_canonical() {
        // Canonical IPv4 -- not ambiguous (use is_ip_literal).
        assert!(!is_ambiguous_ip("127.0.0.1"));
        assert!(!is_ambiguous_ip("169.254.169.254"));
        assert!(!is_ambiguous_ip("10.0.0.1"));

        // IPv6 -- not ambiguous.
        assert!(!is_ambiguous_ip("::1"));
        assert!(!is_ambiguous_ip("fe80::1"));
        assert!(!is_ambiguous_ip("[::1]"));
        assert!(!is_ambiguous_ip("[::ffff:127.0.0.1]"));

        // Genuine hostnames -- not ambiguous.
        assert!(!is_ambiguous_ip("example.com"));
        assert!(!is_ambiguous_ip("localhost"));
        assert!(!is_ambiguous_ip("metadata.google.internal"));
        assert!(!is_ambiguous_ip("api.example.com"));

        // Empty / invalid -- not ambiguous.
        assert!(!is_ambiguous_ip(""));
        assert!(!is_ambiguous_ip("not-an-ip"));
    }
}

File: go-native/ressrf/internal/engine/cidr.go
package engine

import (
	"net/netip"
	"strconv"
	"strings"
)

// CIDR is a parsed CIDR range tagged with the tier it came from.
//
// The underlying matcher uses netip.Prefix in its canonical (Masked) form so
// containment checks are allocation-free and handle IPv4-mapped IPv6 the same
// way as the Rust core (which normalizes everything to ::ffff:0:0/96-relative).
type CIDR struct {
	Prefix   netip.Prefix
	Original string
	Source   DataTier
}

func (c CIDR) String() string { return c.Original }

// ParseCIDR is the strict parser used for built-in / IANA / cloud-module
// entries. It rejects: octal IPv4 octets, hex IPv4 octets, IPv6 zone IDs, and
// host bits set in the network portion.
func ParseCIDR(s string, source DataTier) (CIDR, error) {
	return parseCIDR(s, source, true)
}

// ParseCIDRLoose is the user-facing parser: same as ParseCIDR but it accepts
// host bits and silently masks them to the network address.
func ParseCIDRLoose(s string, source DataTier) (CIDR, error) {
	return parseCIDR(s, source, false)
}

func parseCIDR(s string, source DataTier, strict bool) (CIDR, error) {
	s = strings.TrimSpace(s)
	if strings.Contains(s, "%") {
		return CIDR{}, &ParseError{Detail: "zone IDs not allowed in CIDR notation"}
	}
	slash := strings.LastIndexByte(s, '/')
	if slash < 0 {
		return CIDR{}, &ParseError{Detail: "missing '/' in CIDR notation"}
	}
	addrStr, prefixStr := s[:slash], s[slash+1:]

	prefixLen64, err := strconv.ParseUint(prefixStr, 10, 32)
	if err != nil {
		return CIDR{}, &ParseError{Detail: "invalid prefix length"}
	}
	prefixLen := int(prefixLen64)

	addr, perr := parseIPStrict(addrStr)
	if perr != nil {
		return CIDR{}, perr
	}

	maxPrefix := 128
	if addr.Is4() {
		maxPrefix = 32
	}
	if prefixLen > maxPrefix {
		return CIDR{}, &ParseError{Detail: "prefix length exceeds address family maximum"}
	}

	prefix, perr2 := netip.AddrFrom16(addr.As16()).Prefix(addrPrefixBits(addr, prefixLen))
	if perr2 != nil {
		return CIDR{}, &ParseError{Detail: "invalid prefix: " + perr2.Error()}
	}
	if strict {
		if prefix.Masked() != prefix {
			return CIDR{}, &ParseError{Detail: "host bits set in network address"}
		}
		return CIDR{Prefix: prefix, Original: s, Source: source}, nil
	}
	return CIDR{Prefix: prefix.Masked(), Original: s, Source: source}, nil
}

// Contains reports whether the given address falls inside this CIDR.
func (c CIDR) Contains(addr netip.Addr) bool {
	if !addr.IsValid() {
		return false
	}
	return c.Prefix.Contains(netip.AddrFrom16(addr.As16()))
}

// NativePrefix returns the prefix in native form: an IPv4 range is returned as
// a 4-byte (Is4) prefix rather than the IPv4-mapped IPv6 form the matcher
// stores internally (a.b.c.d/n instead of ::ffff:a.b.c.d/(n+96)). IPv6 ranges
// are returned unchanged. Use this when handing the range to code that expects
// human / family-native CIDRs, e.g. rendering a Kubernetes NetworkPolicy
// except-list.
func (c CIDR) NativePrefix() netip.Prefix {
	addr := c.Prefix.Addr()
	if addr.Is4In6() {
		return netip.PrefixFrom(addr.Unmap(), c.Prefix.Bits()-96)
	}
	return c.Prefix
}

// CIDRSet is a flat collection of CIDRs scanned in insertion order. The Rust
// core uses the same linear scan; for ~30 entries it outperforms tree
// structures because of cache locality.
type CIDRSet struct {
	Ranges []CIDR
}

func (s *CIDRSet) Add(c CIDR) { s.Ranges = append(s.Ranges, c) }

// Contains returns the first CIDR in the set that matches the address, or nil.
func (s *CIDRSet) Contains(addr netip.Addr) *CIDR {
	for i := range s.Ranges {
		if s.Ranges[i].Contains(addr) {
			return &s.Ranges[i]
		}
	}
	return nil
}

func (s *CIDRSet) Len() int      { return len(s.Ranges) }
func (s *CIDRSet) IsEmpty() bool { return len(s.Ranges) == 0 }

// addrPrefixBits returns the prefix length in the IPv6 address space: IPv4
// addresses are normalized to ::ffff:a.b.c.d so their prefix shifts by 96.
func addrPrefixBits(addr netip.Addr, prefixLen int) int {
	if addr.Is4() {
		return prefixLen + 96
	}
	return prefixLen
}

// parseIPStrict parses an IP literal while rejecting Rust-style ambiguous
// IPv4 forms (octal, hex, shorthand). It does not strip brackets.
func parseIPStrict(s string) (netip.Addr, error) {
	s = strings.TrimSpace(s)
	if strings.Contains(s, ":") {
		// Rust's stdlib Ipv6Addr::from_str rejects RFC 6874 zone identifiers
		// (the `%scope` suffix). Go's netip.ParseAddr accepts them. Match
		// Rust's stricter behavior so the differential harness sees identical
		// IP-literal decisions on both sides; URLs that look like IPv6+zone
		// fall through to hostname treatment in both engines.
		if strings.Contains(s, "%") {
			return netip.Addr{}, &ParseError{Detail: "zone IDs not allowed in IP address"}
		}
		addr, err := netip.ParseAddr(s)
		if err != nil {
			return netip.Addr{}, &ParseError{Detail: "invalid IPv6 address: " + err.Error()}
		}
		return addr, nil
	}
	if err := rejectAmbiguousIPv4(s); err != nil {
		return netip.Addr{}, err
	}
	addr, err := netip.ParseAddr(s)
	if err != nil {
		return netip.Addr{}, &ParseError{Detail: "invalid IPv4 address: " + err.Error()}
	}
	if !addr.Is4() {
		return netip.Addr{}, &ParseError{Detail: "expected IPv4 address"}
	}
	return addr, nil
}

func rejectAmbiguousIPv4(s string) error {
	for octet := range strings.SplitSeq(s, ".") {
		if octet == "" {
			return &ParseError{Detail: "empty octet in IPv4 address"}
		}
		if strings.HasPrefix(octet, "0x") || strings.HasPrefix(octet, "0X") {
			return &ParseError{Detail: "hex notation not allowed in IP addresses"}
		}
		if len(octet) > 1 && octet[0] == '0' {
			return &ParseError{Detail: "leading zeros (octal notation) not allowed in IP addresses"}
		}
	}
	return nil
}

// ParseIP exposes the strict parser for use by the URI validator.
func ParseIP(s string) (netip.Addr, error) { return parseIPStrict(s) }

// IsIPLiteral reports whether s looks like an unambiguous IP literal, with
// optional brackets. Ambiguous forms (octal, hex, decimal, shorthand) return
// false here; call IsAmbiguousIP separately to detect those.
func IsIPLiteral(s string) bool {
	s = strings.TrimSpace(s)
	inner := s
	if strings.HasPrefix(s, "[") && strings.HasSuffix(s, "]") {
		inner = s[1 : len(s)-1]
	}
	_, err := parseIPStrict(inner)
	return err == nil
}

// IsAmbiguousIP detects hosts that look like IPs but use non-canonical
// encodings (decimal integer, octal, hex, shorthand). These must be rejected
// before reaching DNS to prevent silent fall-through to internal resolvers.
func IsAmbiguousIP(s string) bool {
	s = strings.TrimSpace(s)
	if s == "" {
		return false
	}
	// Bracketed = IPv6, unambiguous (parses or doesn't).
	if strings.HasPrefix(s, "[") && strings.HasSuffix(s, "]") {
		return false
	}
	// IPv6 with colons: unambiguous.
	if strings.Contains(s, ":") {
		return false
	}
	// Canonical 4-octet IPv4: unambiguous.
	if _, err := parseIPStrict(s); err == nil {
		return false
	}
	lower := strings.ToLower(s)
	if isAmbiguousHexLabel(lower) {
		return true
	}
	for label := range strings.SplitSeq(lower, ".") {
		if isAmbiguousHexLabel(label) {
			return true
		}
	}
	if allDigits(s) {
		return true
	}
	if strings.Contains(s, ".") {
		allNum := true
		hasOctal := false
		octetCount := 0
		for label := range strings.SplitSeq(s, ".") {
			octetCount++
			if label == "" {
				allNum = false
				break
			}
			if !allDigits(label) {
				allNum = false
				break
			}
			if len(label) > 1 && label[0] == '0' {
				hasOctal = true
			}
		}
		if allNum {
			if hasOctal {
				return true
			}
			if octetCount != 4 {
				return true
			}
		}
	}
	return false
}

func allDigits(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// isAmbiguousHexLabel mirrors Rust's `lower.starts_with("0x") &&
// trim_start_matches("0x").chars().all(is_ascii_hexdigit)`. The Rust
// trim strips ALL leading "0x" pairs (so "0x0x0" -> "0"), and `.all()`
// on an empty tail is vacuously true (so bare "0x" is itself ambiguous).
// Caller must pass an already-lowercased string.
func isAmbiguousHexLabel(s string) bool {
	if !strings.HasPrefix(s, "0x") {
		return false
	}
	tail := s
	for strings.HasPrefix(tail, "0x") {
		tail = tail[2:]
	}
	return tail == "" || allHexDigits(tail)
}

func allHexDigits(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c >= '0' && c <= '9':
		case c >= 'a' && c <= 'f':
		case c >= 'A' && c <= 'F':
		default:
			return false
		}
	}
	return true
}
